Tests/SwiftFusionTests/Core/VectorTests.swift

```diff
@@ -1,292 +1,193 @@
-      XCTAssertEqual(pb(v).1, v)
-    }
-  }
-
-  /// Tests the value and derivative of the mutating -=.
-  func testSubtractMutating() {
-    func doMutatingSubtract(_ v1: Testee, _ v2: Testee) -> Testee {
-      var result = v1
-      result -= v2
-      return result
-    }
-
-    let v1 = makeVector(from: 1, stride: 1)
-    let v2 = makeVector(from: 1, stride: -1)
-    XCTAssertEqual(doMutatingSubtract(v1, v2), makeVector(from: 0, stride: 2))
-
-    let (value, pb) = valueWithPullback(at: v1, v2, in: doMutatingSubtract)
-    XCTAssertEqual(value, makeVector(from: 0, stride: 2))
-    for v in basisVectors {
-      XCTAssertEqual(pb(v).0, v)
-      XCTAssertEqual(pb(v).1, -v)
-    }
-  }
-
-  /// Tests the value and derivative of the functional -.
-  func testSubtractFunctional() {
-    let v1 = makeVector(from: 1, stride: 1)
-    let v2 = makeVector(from: 1, stride: -1)
-    XCTAssertEqual(v1 - v2, makeVector(from: 0, stride: 2))
-
-    let (value, pb) = valueWithPullback(at: v1, v2, in: -)
-    XCTAssertEqual(value, makeVector(from: 0, stride: 2))
-    for v in basisVectors {
-      XCTAssertEqual(pb(v).0, v)
-      XCTAssertEqual(pb(v).1, -v)
+extension Vector {
+  /// XCTests `self`'s semantic conformance to `Vector`, expecting its scalars to match
+  /// `expectedScalars`.
+  ///
+  /// - Parameter distinctScalars: scalars that are elementwise distinct from `expectedScalars`.
+  ///   These are written to the scalars collection to test mutable collection semantics.
+  /// - Parameter maxSupportedScalarCount: the maximum number of scalars  that instances of `Self`
+  ///   can have.
+  /// - Requires: `!distinctScalars.elementsEqual(self.scalars)`.
+  /// - Complexity: O(N²), where N is `self.dimension`.
+  func checkVectorSemantics<S1: Collection, S2: Collection>(
+    expectingScalars expectedScalars: S1,
+    writingScalars distinctScalars: S2,
+    maxSupportedScalarCount: Int = Int.max
+  ) where S1.Element == Double, S2.Element == Double {
+
+    // MARK: - Check the `scalars` property semantics.
+
+    self.scalars.checkCollectionSemantics(
+      expecting: expectedScalars, maxSupportedCount: maxSupportedScalarCount)
+
+    var mutableScalars = self.scalars
+    mutableScalars.checkMutableCollectionSemantics(writing: distinctScalars)
+
+    // Check that setting `scalars` actually changes it.
+    var mutableSelf = self
+    mutableSelf.scalars = mutableScalars
+    XCTAssertTrue(mutableSelf.scalars.elementsEqual(mutableScalars))
+
+    // MARK: - Check the `dimension` property semantics.
+
+    XCTAssertEqual(self.dimension, expectedScalars.count)
+
+    // MARK: - Check the mathematical vector operation (`+`, `+=`, `-`, `-=`, `*`, `*=`, `dot`,
+    // `zeroValue`) semantics.
+
+    /// Returns `self` with its scalars replaced with `start`, `start + stride`,
+    /// `start + 2 * stride`, etc.
+    func strideVector(from start: Double, by stride: Double) -> Self {
+      var r = self
+      for (i, e) in zip(
```

Comment:
@marcrasi FYI, we have algorithms for this kind of open-coded loop, now!

See PenguinExtensions.swift extensions on MutableCollection

Reply:
```swift
      r.scalars.assign(
        Swift.stride(from: start, to: start + Double(r.dimension) * stride, by: stride))
```